File: 0018_4Sum/4Sum.cpp
// 3Sum er opor arek ta for loop kora hoyeche
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        vector<vector<int>> ans;

        int n = nums.size();
        sort(nums.begin(), nums.end());
        for(int i = 0; i < n-3; i++){
            if (i > 0 && nums[i]==nums[i-1]) continue;  // Most outer loop.

            for(int j = i+1; j < n-2; j++){
                if (j > (i+1) && nums[j]==nums[j-1]) continue;  // Second loop. If confused about j>i+1, then compare it with i's loop's one

                // Two pointers
                int k = j+1;        
                int l = n-1;
                while (k < l){
                    // Input maybe very large so Long Long
                    // Expressions like (nums[i] + nums[j] + nums[k] + nums[l]) are evaluated in int and can overflow before you compare to target.
                    long long sum = nums[i];  
                    sum += nums[j];
                    sum += nums[k];
                    sum += nums[l];
                    // Some SOAB inputs can be like this, [1000000000,1000000000,1000000000,1000000000]. thats why i used 1 by 1 summation method
                    if (sum < target){
                        k++;
                        //negative condition
                    }
                    else if(sum > target){
                        l--;// positive condition
                    }
                    else{
                        vector<int> temp = {nums[i], nums[j], nums[k], nums[l]};
                        ans.push_back(temp);
                        //zero condition
                        k++;
                        l--;
                        while((k < l) && nums[k] == nums[k-1]) k++;
                        while ((k < l) && nums[l] == nums[l+1]) l--;
                    }
                }
            }
        }
        return ans;
    }
};

File: 0018_4Sum/4Sum.py
# Check Cpp file for explaining comments

class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:

        n = len(nums)
        ans = []
        nums.sort()
        for i in range(0 , n-3):
            if (i > 0 and nums[i]==nums[i-1]): 
                continue

            for j in range(i+1 , n-2):
                if (j > (i+1) and nums[j]==nums[j-1]): 
                    continue

                k = j+1
                l = n-1
                while (k < l):
                    sum = nums[i] + nums[j] + nums[k] + nums[l]

                    if (sum < target): 
                        k+= 1
                        #negative condition
                    elif (sum > target):
                        l-= 1 # positive condition
                    else:
                        temp = [nums[i], nums[j], nums[k], nums[l]]
                        ans.append(temp)
                        #zero condition
                        k+=1
                        l-=1
                        while((k < l) and nums[k] == nums[k-1]):
                            k+=1
                        while ((k < l) and nums[l] == nums[l+1]):
                            l-=1
                            
        return ans